Backports for 1.12.0-rc2

KristofferC opened the pull request:
Backported PRs:
- [x] #57500 <!-- make compile go faster with native caches -->
- [x] #58997 <!-- add array element mutex offset in print and gc -->
- [x] #59112 <!-- Add builtin function name to add methods error -->
- [x] #59120 <!-- Fix memory order typo in "src/julia_atomics.h" -->
- [x] #58650 <!-- fix serializer compat with CodeInfos from v1.11 -->
- [x] #59163 <!-- misc: Don't get confused by GC running during compilation -->
- [x] #59170 <!-- Clarify and enhance confusing precompile test -->
- [x] #59155 <!-- Fix desugaring of `const x::T = y` for complex `y` -->
- [x] #59158 <!-- rename `GlobalMethods` to `methodtable` :bike: :house:  -->
- [x] #59192 <!-- Export jl_declare_constant_val -->
- [x] #58872 <!-- fix #57749, model ccall binding access in inference -->
- [x] #59210 <!-- fix runtime cglobal builtin function implementation -->
- [x] #59259 <!-- Fix compile time regression for null ptr comparisons -->
- [x] #59279 <!-- fix #59252, make `args` in juliac like it is elsewhere -->

Need manual backport:
- [ ] #57410 <!-- codegen: cleanup gcstack call frames somewhat earlier -->
- [ ] #59028 <!-- build: More msys2 fixes -->
- [ ] #59123 <!-- [REPL] Fix keyword arguments completions with do block -->
- [ ] #59164 <!-- build: Fix libssh2 source build -->

Contains multiple commits, manual intervention needed:
- [ ] #57758 <!-- improve sysimage.mk more completely -->
- [ ] #57143 <!-- Add JuliaLang/JuliaSyntax.jl#525 to NEWS.md, flisp parser, and REPL -->

Non-merged PRs with backport label:
- [x] #59299 <!-- Fix printing of errors in juliac + searchsorted functions for LinearAlgebra -->
- [x] #59298 <!-- Add `_assert_to_string` indirection for `Expr`-evaluated `@assert` -->
- [x] #59297 <!-- juliac: apply `buildscript` modifications after user code -->
- [x] #59232 <!-- Make `Base.disable_library_threading_hooks` ephemeral to the process -->
- [ ] #59278 <!-- fix #58780, make `@allocated` more reliable and handle assignment syntax -->
- [ ] #59276 <!-- add `macroexpand!` function and add `legacyscope` kwarg -->
- [ ] #59201 <!-- Move jl_init_ to the public header -->
- [ ] #59165 <!-- ccall: make distinction of pointer vs name a syntactic distinction -->
- [ ] #59111 <!-- Fix a JET warning -->
- [ ] #58535 <!-- gf.c: include const-return methods in `--trace-compile` -->
- [ ] #58359 <!-- Use dwarf for symbol lookup if dladdr fails before doing linear search. -->
- [ ] #58124 <!-- Log a backtrace on entrance to type inference -->
- [ ] #58038 <!-- strings/cstring: `transcode`: prevent Windows sysimage invalidation -->
- [ ] #57604 <!-- `@nospecialize` for `string_index_err` -->
- [ ] #57490 <!-- export `maxthreadid` from `Threads` -->
- [ ] #57422 <!-- Fix performance regression in hvcat of simple matrices -->
- [ ] #57366 <!-- Use ptrdiff_t sized offsets for gvars_offsets to allow large sysimages -->
- [ ] #56823 <!-- Make version of opaque closure constructor in world -->

vtjnash committed:
make compile go faster with native caches (#57500)

Not sure why we make sure compile go slow, so make it go fast.

(cherry picked from commit 095fc99b69459d43c2ab3267ade9f491f52b92cc)

vtjnash committed:
add array element mutex offset in print and gc (#58997)

The layout, printing, and gc logic need to correctly offset and align
the inset fields to account for the per-element mutex of an atomic
array with large elements.

Fix #58993

(cherry picked from commit c5ba4e28f2c9c49be01ed7409efedef8627fe96d)

PallHaraldsson commented:
rc2 presumably?

IanButterworth committed:
Add builtin function name to add methods error (#59112)

```
julia> Base.throw(x::Int) = 1
ERROR: cannot add methods to builtin function `throw`
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```

(cherry picked from commit 57def4d19ef79794439ef839e5e3b551323f49a2)

omus committed:
Fix memory order typo in "src/julia_atomics.h" (#59120)

(cherry picked from commit fde79f5bb5c4835189bfe3215521137176ac3eb8)

IanButterworth committed:
update libblastrampoline_jll to fix julia compat

giordano commented:
@IanButterworth Do you need to change also https://github.com/JuliaLang/julia/blob/0f3c3659fa82a5ea51c8b786848da6be6e7b83ec/stdlib/libblastrampoline_jll/Project.toml#L10? I don't know if that's used anywhere.

JeffBezanson committed:
fix serializer compat with CodeInfos from v1.11 (#58650)

(cherry picked from commit 9f6996149a9b56c03cf1ccff051e35c225539c04)

aviatesk committed:
REPL: improve type inference for `maybe_spawn_cache_PATH` (#59144)

aviatesk committed:
iterator: improve type inference for `Filter` (#59142)

Keno committed:
misc: Don't get confused by GC running during compilation (#59163)

Fixes https://github.com/JuliaLang/julia/issues/59162

(cherry picked from commit f5c55e808cf8c3da77fbcc3db712ff2cb88d2c25)

IanButterworth committed:
Clarify and enhance confusing precompile test (#59170)

(cherry picked from commit 1f6eff183db0b947632f780b4acc440d9c41a6e2)

mlechu committed:
Fix desugaring of `const x::T = y` for complex `y` (#59155)

Fix #59128

Assignment desugaring of `(const (= (|::| x T) rhs))` would pre-expand
to, then re-expand `(const x ,(convert-for-type-decl rhs T))`, but
two-arg (IR) const is expected to have a simple RHS---closure conversion
doesn't recurse here (should it?), giving us partially-lowered IR, and
hence our bug.

Fix: Pre-expand to the one-arg AST const form `(const (= x
,(convert-for-type-decl rhs T)))` instead. This also gives us a
`(latestworld)` we were missing before, so this lowering may have been
originally intended.

(cherry picked from commit 3de5b9a66d83a9b4dab665d47c4e3afcf5388a63)

aviatesk committed:
inference: avoid `LimitedAccuracy` within slot wrappers (#59182)

I looked at all the usage sites of these constructors within the
compiler, and fortunately it appears that none of the usage sites
require that the return values of these constructors be objects of their
respective types. So in cases where `LimitedAccuracy` is given as a
wrapped element, these slot type refinements simply will not be
performed.

- fixes JuliaLang/julia#59004

JeffBezanson committed:
rename `GlobalMethods` to `methodtable` :bike: :house:  (#59158)

(cherry picked from commit f9b7d27b753020b828d6a050a341b3c2af49000c)

Taaitaaiger committed:
Export jl_declare_constant_val (#59192)

Closes #59184

(cherry picked from commit 13f6b3b5c0361c865ba6f3e6f09fb7d3ae6a6735)

JeffBezanson committed:
model ccall binding access in inference (#58872)

fixes #57749

(cherry picked from commit b35c4f471f05873a4de6a989e0038a5c28cd09ce)

topolarity committed:
Revert "remove hand-coded methods for (U)Int128 on 32 bit systems (#53867)" (#59211)

We need to revert this for now. Resolves
https://github.com/JuliaLang/julia/issues/58911

It seems that the back-end LLVM emits this libcall a lot less often than
it used to, but it still happens sometimes and 32-bit Windows has no
implementation of `__divti3`.

vtjnash committed:
fix runtime cglobal builtin function implementation (#59210)

This had failed to be updated for the LazyLibrary changes to codegen.

(cherry picked from commit 46c2a5c7e1f970e83b408b6ddcba49aaa31d8329)

IanButterworth committed:
move threading news to threading section. Add reminder about threadid

gbaraldi committed:
Fix compile time regression for null ptr comparisons (#59259)

Fixes the compile time regression in
https://github.com/JuliaLang/julia/issues/59134 (does not address the
performance regression that should be tracked and bisected)

(cherry picked from commit 920df7abc9b5bf47d0589d9768b41e41421ad683)

KristofferC commented:
```

ccall                                             (1) \|         failed at 2025-08-13T13:07:02.330
--
  | Error During Test at /cache/build/tester-amdci5-13/julialang/julia-release-1-dot-12/julia-721648259c/share/julia/test/testdefs.jl:24
  | Got exception outside of a @test
  | LoadError: ArgumentError: Package Zstd_jll not found in current path.
  | - Run `import Pkg; Pkg.add("Zstd_jll")` to install the Zstd_jll package.
  | Stacktrace:
  | [1] macro expansion
  | @ ./loading.jl:2371 [inlined]
  | [2] macro expansion
  | @ ./lock.jl:376 [inlined]
  | [3] __require(into::Module, mod::Symbol)
  | @ Base ./loading.jl:2354
  | [4] require(into::Module, mod::Symbol)
  | @ Base ./loading.jl:2330
  | [5] top-level scope
  | @ /cache/build/tester-amdci5-13/julialang/julia-release-1-dot-12/julia-721648259c/share/julia/test/ccall.jl:1984
```

Needs to be modified on 1.12 since Zstd_jll is not an stdlib.

JeffBezanson committed:
fix #59252, make `args` in juliac like it is elsewhere (#59279)

Also set PROGRAM_FILE and ARGS.

fixes #59252

(cherry picked from commit 7b9f53231296ea8a368ec802b13d83df04a21802)

topolarity committed:
juliac: apply `buildscript` modifications after user code (#59297)

These modifications to the MethodTable should be performed at the
"link-stage", after all user top-level code has executed. This should be
closer to the upcoming 1.13 pipeline, which will move all top-level user
code to its standard package execution context.

This also makes any user-provided functionality for type printing
available to the `TrimVerifier`.

Resolves https://github.com/JuliaLang/julia/issues/59280.

topolarity committed:
Add `_assert_to_string` indirection for `Expr`-evaluated `@assert` (#59298)

Not sure why this wasn't using this before since the code is effectively
identical, but this allows:
```julia
@assert false "$("Hello,") assertion!"
```
to be `--trim`-"compatible" (unfortunately by deleting the assertion
message...)

gbaraldi committed:
Fix printing of errors in juliac + searchsorted functions for LinearAlgebra (#59299)

We may want to move this changes to base proper.
@LilithHafner should sort functions just always specialize on the
`by`/`lt` params? It currently sometimes relies on inlining for good
performance

---------

Co-authored-by: Cody Tapscott <84105208+topolarity@users.noreply.github.com>
Co-authored-by: Cody Tapscott <topolarity@tapscott.me>

gbaraldi committed:
Use Zlib instead of zstd

topolarity committed:
Make `Base.disable_library_threading_hooks` ephemeral to the process (#59232)

This prevents accidentally registering hooks during the sysimage build
that would persist to run-time (unlike what happens when registering
hooks while building a pkgimage)

~~Required for https://github.com/JuliaLang/LinearAlgebra.jl/pull/1407~~
